File: AccuPay/Domain/Payroll/CashOutUnusedLeave.vb
Imports System.Threading.Tasks
Imports AccuPay.Data.Repositories
Imports AccuPay.Data.Services
Imports AccuPay.Entity
Imports Microsoft.EntityFrameworkCore

Public Class CashOutUnusedLeave

    Private Const _defaultWorkHours As Decimal = 8.0

    Private _settings As ListOfValueCollection

    'Private _adjUnusedVacationLeave As String = String.Join(Space(1), "Unused", LeaveType.Vacation.ToString(), "Leaves")
    Private _adjUnusedVacationLeave As String = String.Join(Space(1), "Unused", "Leaves")

    Private _adjUnusedSickLeave As String = String.Join(Space(1), "Unused", LeaveType.Sick.ToString(), "Leaves")

    Private _isVLOnly, _isSLOnly, _asAdjustment As Boolean

    Private _payPeriodFromId,
        _payPeriodToId,
        _adjUnusedVacationLeaveId,
        _adjUnusedSickLeaveId,
        _currentPeriodId,
        _organizationId,
        _vacationLeaveId,
        _sickLeaveId As Integer

    Private _leaveLedger As DataTable
    Private _categoryRepository As CategoryRepository
    Private Const strAdjType = "Adjustment Type"

    Private Const strLeaveType = "Leave Type"

    Private _employeeRepository As EmployeeRepository

    Private _listOfValRepository As ListOfValueRepository

    Public Sub New(PayPeriodFromId As Integer,
                   PayPeriodToId As Integer,
                   CurrentPeriodID As Integer)

        _categoryRepository = New CategoryRepository

        _employeeRepository = New EmployeeRepository

        _organizationId = Convert.ToInt32(z_OrganizationID)

        Dim listOfValues = _listOfValRepository.GetLeaveConvertiblePolicies()

        Dim _leaveType = listOfValues.Where(Function(l) Equals(l.LIC, "LeaveType")).FirstOrDefault

        _isVLOnly = _leaveType.DisplayValue = LeaveType.Vacation.ToString()

        _isSLOnly = _leaveType.DisplayValue = LeaveType.Sick.ToString()

        Dim _amountTreatment = listOfValues.Where(Function(l) Equals(l.LIC, "AmountTreatment")).FirstOrDefault

        _asAdjustment = _amountTreatment.DisplayValue = AmountTreatment.Adjustment.ToString()

        'Dim policy = _settings.GetSublist("LeaveConvertiblePolicy")

        '_isVLOnly = policy.GetValue("LeaveType") = LeaveType.Vacation.ToString()

        '_isSLOnly = policy.GetValue("LeaveType") = LeaveType.Sick.ToString()

        '_asAdjustment = policy.GetValue("AmountTreatment") = AmountTreatment.Adjustment.ToString()

        _payPeriodFromId = PayPeriodFromId
        _payPeriodToId = PayPeriodToId
        _currentPeriodId = CurrentPeriodID

        LoadAdjUnusedVacationLeaveId()

        LoadVacationLeaveId()

    End Sub

    Private Async Sub LoadVacationLeaveId()
        _vacationLeaveId = Await GetVacationLeaveId()
    End Sub

    Private Async Sub LoadAdjUnusedVacationLeaveId()
        _adjUnusedVacationLeaveId = Await GetAdjUnusedVacationLeaveId()
    End Sub

    Private Function GetLatestLeaveLedger() As DataTable
        Dim query1 = $"CALL RPT_LeaveConvertibles(@orgId, @leaveTypeId, @payPeriodFromId, @payPeriodToId, NULL);"

        Dim leaveTypeId? As Integer = 0

        Dim params = New Dictionary(Of String, Object) From {
            {"@orgId", _organizationId},
            {"@leaveTypeId", If(_isVLOnly, _vacationLeaveId, DBNull.Value)},
            {"@payPeriodFromId", _payPeriodFromId},
            {"@payPeriodToId", _payPeriodToId}
        }

        Dim query = New SqlToDataTable(query1, params)

        Return query.Read()
    End Function

    Private Async Function GetVacationLeaveId() As Threading.Tasks.Task(Of Integer)
        Dim value As Integer
        Dim strVacationLeave = String.Join(Space(1), LeaveType.Vacation.ToString(), "Leave")

        Dim leaveCategory = Await _categoryRepository.GetByNameAsync(z_OrganizationID, strVacationLeave)
        Dim leaveCategoryId = leaveCategory.RowID

        Using context = New PayrollContext()

            Dim product = context.Products.
                Where(Function(p) Equals(p.PartNo.ToLower, strVacationLeave.ToLower) _
                       And Equals(p.OrganizationID.Value, _organizationId) _
                       And Equals(p.CategoryID, leaveCategoryId) _
                       And Equals(p.Category, strLeaveType)).FirstOrDefault

            value = product.RowID
        End Using

        Return value
    End Function

    Private Async Function GetAdjUnusedVacationLeaveId() As Threading.Tasks.Task(Of Integer)
        Dim value As Integer

        Dim adjustmentCategory = Await _categoryRepository.GetByNameAsync(z_OrganizationID, strAdjType)
        Dim adjustmentCategoryId = adjustmentCategory.RowID

        Using context = New PayrollContext()
            Dim product = context.Products.
                Where(Function(p) Equals(p.PartNo.ToLower, _adjUnusedVacationLeave.ToLower) _
                       And Equals(p.OrganizationID.Value, _organizationId) _
                       And Equals(p.CategoryID, adjustmentCategoryId) _
                       And Equals(p.Category, strAdjType)).FirstOrDefault

            If product Is Nothing Then
                value = CreateProduct(context, _adjUnusedVacationLeave, strAdjType, adjustmentCategoryId)
            Else
                value = product.RowID
            End If

        End Using

        Return value
    End Function

    Private Async Function GetAdjUnusedSickLeaveId() As Threading.Tasks.Task(Of Integer)
        Dim value As Integer

        Dim adjustmentCategory = Await _categoryRepository.GetByNameAsync(z_OrganizationID, strAdjType)
        Dim adjustmentCategoryId = adjustmentCategory.RowID

        Using context = New PayrollContext()
            Dim product = context.Products.
                Where(Function(p) Equals(p.PartNo.ToLower, _adjUnusedSickLeave.ToLower) _
                       And Equals(p.OrganizationID.Value, _organizationId) _
                       And Equals(p.CategoryID, adjustmentCategoryId) _
                       And Equals(p.Category, strAdjType)).FirstOrDefault

            If product Is Nothing Then
                value = CreateProduct(context, _adjUnusedSickLeave, strAdjType, adjustmentCategoryId)
            Else
                value = product.RowID
            End If

        End Using

        Return value
    End Function

    Private Function CreateProduct(context As PayrollContext, ProductName As String, CategoryName As String, CategoryRowId As Integer) As Integer
        Dim product As New Product With {
            .Name = ProductName,
            .OrganizationID = _organizationId,
            .Description = String.Empty,
            .PartNo = ProductName,
            .Created = Date.Now,
            .LastUpd = Date.Now,
            .CreatedBy = z_User,
            .LastUpdBy = z_User,
            .Category = CategoryName,
            .CategoryID = CategoryRowId,
            .Status = "Active",
            .Fixed = False
        }

        context.Products.Add(product)

        context.SaveChanges()

        Return product.RowID
    End Function

    Public Sub Execute()
        _leaveLedger = GetLatestLeaveLedger()

        Dim payperiod As New PayPeriod

        If _isVLOnly And _asAdjustment Then

            Dim success As Boolean = False

            Using context = New PayrollContext()
                Dim paystubs =
                    (From p In context.Paystubs.Include(Function(p) p.ActualAdjustments)
                     Where p.PayPeriodID.Value = _currentPeriodId And p.OrganizationID = _organizationId).
                     ToList()

                payperiod = context.PayPeriods.Where(Function(pp) Equals(pp.RowID.Value, _currentPeriodId)).FirstOrDefault

                For Each row As DataRow In _leaveLedger.Rows
                    Dim employeePrimKey = Convert.ToInt32(row("EmployeeID"))
                    Dim llRowId = Convert.ToInt32(row("RowID"))

                    Dim ll As New LeaveLedger With {.RowID = llRowId, .EmployeeID = employeePrimKey}

                    Dim unusedLeaveAmount = Convert.ToDecimal(row("Balance")) * Convert.ToDecimal(row("HourlyRate"))

                    Dim leaveBalance = Convert.ToDecimal(row("Balance"))
                    Dim leaveDayBalance = leaveBalance / _defaultWorkHours

                    Dim paystub = paystubs.Where(Function(p) Equals(p.EmployeeID.Value, employeePrimKey)).FirstOrDefault

                    If paystub IsNot Nothing Then
                        Dim adjustmentsExceptUnusedLeave =
                            paystub.ActualAdjustments.Where(Function(a) Equals(a.ProductID.Value, _adjUnusedVacationLeaveId) = False)

                        paystub.ActualAdjustments.Clear()

                        Dim aa As New ActualAdjustment With {
                        .Created = Date.Now,
                        .Comment = String.Concat(leaveDayBalance, If(leaveDayBalance > 1, " days", " day")),
                        .CreatedBy = z_User,
                        .IsActual = True,
                        .LastUpd = Date.Now,
                        .LastUpdBy = z_User,
                        .OrganizationID = _organizationId,
                        .PayAmount = unusedLeaveAmount,
                        .PayStubID = paystub.RowID,
                        .ProductID = _adjUnusedVacationLeaveId
                        }

                        paystub.ActualAdjustments.Add(aa)

                        For Each adj In adjustmentsExceptUnusedLeave
                            paystub.ActualAdjustments.Add(adj)
                        Next

                        CreateLeaveTransaction(context, LeaveTransactionType.Debit, ll, payperiod, leaveBalance)
                    Else

                        Continue For
                    End If
                Next

                context.SaveChanges()
                success = True
            End Using

            If success Then
                Dim strCutOff As String
                Using context = New PayrollContext()
                    If payperiod Is Nothing Then
                        payperiod = context.PayPeriods.Where(Function(pp) Equals(pp.RowID.Value, _currentPeriodId)).FirstOrDefault
                    End If

                    strCutOff = String.Join(" to ", payperiod.PayFromDate.ToShortDateString, payperiod.PayToDate.ToShortDateString)
                End Using
                MsgBox(String.Concat("Unused leaves were successfully computed.", vbNewLine, "Please generate the ", strCutOff, " payroll."), MsgBoxStyle.Information)
            End If
        Else

        End If

    End Sub

    Private Async Sub ZeroOutEmployeeLeaveBalance(context As PayrollContext, employeeRowId As Integer)

        Dim employee = Await _employeeRepository.GetByIdAsync(employeeRowId)

        If employee IsNot Nothing Then
            With employee
                .LeaveBalance = 0
                .SickLeaveBalance = 0
            End With

        End If

    End Sub

    Private Function CreateLeaveTransaction(context As PayrollContext, leaveTransactionType As LeaveTransactionType, leaveLedger As LeaveLedger, payPeriod As PayPeriod, unusedLeaveHours As Decimal) As LeaveTransaction
        Dim employeeRowId = leaveLedger.EmployeeID

        Dim lt = context.LeaveTransactions.
            Where(Function(lTrans) Nullable.Equals(lTrans.OrganizationID, z_OrganizationID)).
            Where(Function(lTrans) Nullable.Equals(lTrans.EmployeeID, employeeRowId)).
            Where(Function(lTrans) Nullable.Equals(lTrans.LeaveLedgerID, leaveLedger.RowID)).
            Where(Function(lTrans) Nullable.Equals(lTrans.PayPeriodID, payPeriod.RowID)).
            Where(Function(lTrans) Equals(lTrans.Type, leaveTransactionType.ToString())).
            Where(Function(lTrans) Equals(lTrans.Balance, 0)).
            Where(Function(lTrans) Equals(lTrans.Amount, unusedLeaveHours)).
            Where(Function(lTrans) Equals(lTrans.TransactionDate, payPeriod.PayToDate)).
            FirstOrDefault

        If lt Is Nothing Then
            lt = New LeaveTransaction With {
                .OrganizationID = z_OrganizationID,
                .Created = Date.Now,
                .CreatedBy = z_User,
                .LastUpd = Date.Now,
                .LastUpdBy = z_User,
                .EmployeeID = employeeRowId,
                .LeaveLedgerID = leaveLedger.RowID,
                .PayPeriodID = payPeriod.RowID,
                .ReferenceID = Nothing,
                .TransactionDate = payPeriod.PayToDate,
                .Type = leaveTransactionType.ToString(),
                .Balance = 0,
                .Amount = unusedLeaveHours
                }

            context.LeaveTransactions.Add(lt)

            ZeroOutEmployeeLeaveBalance(context, employeeRowId)

        End If

        Return lt
    End Function

    Enum LeaveTransactionType
        Credit
        Debit
    End Enum

    Enum LeaveType
        Vacation
        Sick
        All
    End Enum

    Enum AmountTreatment
        Adjustment
        Gross
    End Enum

End Class

File: Accupay.Data/Services/CashOutUnusedLeave.cs
using AccuPay.Data.Entities;
using AccuPay.Data.Repositories;
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Threading.Tasks;

namespace AccuPay.Data.Services
{
    /// <summary>
    /// Not working yet
    /// </summary>
    public class CashOutUnusedLeave
    {
        private const decimal _defaultWorkHours = 8.0M;

        // Private _adjUnusedVacationLeave As String = String.Join(Space(1), "Unused", LeaveType.Vacation.ToString(), "Leaves")
        private string _adjUnusedVacationLeave = "Unused Leaves";

        private string _adjUnusedSickLeave = $"Unused {LeaveType.Sick.ToString()} Leaves";

        private bool _isVLOnly, _isSLOnly, _asAdjustment;

        private int _adjUnusedVacationLeaveId, _currentPeriodId, _vacationLeaveId;
        private readonly int _payPeriodFromId;
        private readonly int _payPeriodToId;
        private readonly int _organizationId;
        private readonly int _userId;
        private DataTable _leaveLedger;
        private CategoryRepository _categoryRepository;
        private const string strAdjType = "Adjustment Type";

        private const string strLeaveType = "Leave Type";

        private EmployeeRepository _employeeRepository;

        private ListOfValueRepository _listOfValRepository;

        private PayPeriodRepository _payPeriodRepository;

        public CashOutUnusedLeave(int PayPeriodFromId, int PayPeriodToId, int CurrentPeriodID, int organizationId, int userId)
        {
            _payPeriodFromId = PayPeriodFromId;
            _payPeriodToId = PayPeriodToId;
            _currentPeriodId = CurrentPeriodID;
            _organizationId = organizationId;
            _userId = userId;

            _categoryRepository = new CategoryRepository();

            _employeeRepository = new EmployeeRepository();

            _listOfValRepository = new ListOfValueRepository();

            _payPeriodRepository = new PayPeriodRepository();

            var listOfValues = _listOfValRepository.GetLeaveConvertiblePolicies();

            var _leaveType = listOfValues.Where(l => l.LIC == "LeaveType").FirstOrDefault();

            _isVLOnly = _leaveType.DisplayValue == LeaveType.Vacation.ToString();

            _isSLOnly = _leaveType.DisplayValue == LeaveType.Sick.ToString();

            var _amountTreatment = listOfValues.Where(l => l.LIC == "AmountTreatment").FirstOrDefault();

            _asAdjustment = _amountTreatment.DisplayValue == AmountTreatment.Adjustment.ToString();

            // Dim policy = _settings.GetSublist("LeaveConvertiblePolicy")

            // _isVLOnly = policy.GetValue("LeaveType") = LeaveType.Vacation.ToString()

            // _isSLOnly = policy.GetValue("LeaveType") = LeaveType.Sick.ToString()

            // _asAdjustment = policy.GetValue("AmountTreatment") = AmountTreatment.Adjustment.ToString()

            LoadAdjUnusedVacationLeaveId();

            LoadVacationLeaveId();
        }

        private async void LoadVacationLeaveId()
        {
            _vacationLeaveId = await GetVacationLeaveId();
        }

        private async void LoadAdjUnusedVacationLeaveId()
        {
            _adjUnusedVacationLeaveId = await GetAdjUnusedVacationLeaveId();
        }

        private DataTable GetLatestLeaveLedger()
        {
            return null;
            //var query1 = $"CALL RPT_LeaveConvertibles(@orgId, @leaveTypeId, @payPeriodFromId, @payPeriodToId, NULL);";

            //int? leaveTypeId = 0;

            //var @params = new Dictionary<string, object>()
            //{
            //    {"@orgId", _organizationId},
            //    {"@leaveTypeId", (_isVLOnly ? _vacationLeaveId : DBNull.Value)
            //    },
            //    {"@payPeriodFromId",_payPeriodFromId},
            //    {"@payPeriodToId",_payPeriodToId}
            //};

            //var query = new SqlToDataTable(query1, @params);

            //return query.Read();
        }

        private async Task<int> GetVacationLeaveId()
        {
            // use product repository for this
            int value;
            var strVacationLeave = $"{LeaveType.Vacation.ToString()} Leave";

            var leaveCategory = await _categoryRepository.GetByNameAsync(_organizationId, strVacationLeave);
            var leaveCategoryId = leaveCategory.RowID;

            using (var context = new PayrollContext())
            {
                var product = context.Products.
                            Where(p => p.PartNo.ToLower() == strVacationLeave.ToLower()).
                            Where(p => p.OrganizationID == _organizationId).
                            Where(p => p.CategoryID == leaveCategoryId).
                            Where(p => p.Category == strLeaveType).
                            FirstOrDefault();

                value = product.RowID.Value;
            }

            return value;
        }

        private async Task<int> GetAdjUnusedVacationLeaveId()
        {
            // use product repository for this
            int value;

            var adjustmentCategory = await _categoryRepository.GetByNameAsync(_organizationId, strAdjType);
            var adjustmentCategoryId = adjustmentCategory.RowID;

            using (var context = new PayrollContext())
            {
                var product = context.Products.Where(p => p.PartNo.ToLower() == _adjUnusedVacationLeave.ToLower()).
                                    Where(p => p.OrganizationID == _organizationId).
                                    Where(p => p.CategoryID == adjustmentCategoryId).
                                    Where(p => p.Category == strAdjType).
                                    FirstOrDefault();

                if (product == null)
                    value = CreateProduct(context, _adjUnusedVacationLeave, strAdjType, adjustmentCategoryId);
                else
                    value = product.RowID.Value;
            }

            return value;
        }

        private int CreateProduct(PayrollContext context, string ProductName, string CategoryName, int CategoryRowId)
        {
            Product product = new Product()
            {
                Name = ProductName,
                OrganizationID = _organizationId,
                Description = string.Empty,
                PartNo = ProductName,
                Created = DateTime.Now,
                LastUpd = DateTime.Now,
                CreatedBy = _userId,
                LastUpdBy = _userId,
                Category = CategoryName,
                CategoryID = CategoryRowId,
                Status = "Active",
                Fixed = false
            };

            context.Products.Add(product);

            context.SaveChanges();

            return product.RowID.Value;
        }

        public void Execute()
        {
            _leaveLedger = GetLatestLeaveLedger();

            PayPeriod payperiod = new PayPeriod();

            if (_isVLOnly & _asAdjustment)
            {
                bool success = false;

                payperiod = _payPeriodRepository.GetById(_currentPeriodId);

                using (var context = new PayrollContext())
                {
                    var paystubs = (from p in context.Paystubs.Include(p => p.ActualAdjustments)
                                    where p.PayPeriodID.Value == _currentPeriodId & p.OrganizationID == _organizationId
                                    select p).ToList();

                    foreach (DataRow row in _leaveLedger.Rows)
                    {
                        var employeePrimKey = Convert.ToInt32(row["EmployeeID"]);
                        var llRowId = Convert.ToInt32(row["RowID"]);

                        LeaveLedger ll = new LeaveLedger() { RowID = llRowId, EmployeeID = employeePrimKey };

                        var unusedLeaveAmount = Convert.ToDecimal(row["Balance"]) * Convert.ToDecimal(row["HourlyRate"]);

                        var leaveBalance = Convert.ToDecimal(row["Balance"]);
                        var leaveDayBalance = leaveBalance / _defaultWorkHours;

                        var paystub = paystubs.Where(p => p.EmployeeID.Value == employeePrimKey).FirstOrDefault();

                        if (paystub != null)
                        {
                            var adjustmentsExceptUnusedLeave = paystub.ActualAdjustments.
                                                                Where(a => a.ProductID != _adjUnusedVacationLeaveId);

                            paystub.ActualAdjustments.Clear();

                            ActualAdjustment aa = new ActualAdjustment()
                            {
                                Created = DateTime.Now,
                                Comment = string.Concat(leaveDayBalance, leaveDayBalance > 1 ? " days" : " day"),
                                CreatedBy = _userId,
                                IsActual = true,
                                LastUpd = DateTime.Now,
                                LastUpdBy = _userId,
                                OrganizationID = _organizationId,
                                Amount = unusedLeaveAmount,
                                PaystubID = paystub.RowID,
                                ProductID = _adjUnusedVacationLeaveId
                            };

                            paystub.ActualAdjustments.Add(aa);

                            foreach (var adj in adjustmentsExceptUnusedLeave)
                                paystub.ActualAdjustments.Add(adj);

                            CreateLeaveTransaction(context, LeaveTransactionType.Debit, ll, payperiod, leaveBalance);
                        }
                        else
                            continue;
                    }

                    context.SaveChanges();
                    success = true;
                }

                if (success)
                {
                    string strCutOff;
                    using (var context = new PayrollContext())
                    {
                        if (payperiod == null)
                            payperiod = _payPeriodRepository.GetById(_currentPeriodId);

                        strCutOff = string.Join(" to ", payperiod.PayFromDate.ToShortDateString(), payperiod.PayToDate.ToShortDateString());
                    }
                    // Do this. Was not migrated from VB to C#
                    //Interaction.MsgBox(string.Concat("Unused leaves were successfully computed.", Constants.vbNewLine, "Please generate the ", strCutOff, " payroll."), MsgBoxStyle.Information);
                }
            }
            else
            {
            }
        }

        private async void ZeroOutEmployeeLeaveBalance(int employeeRowId)
        {
            var employee = await _employeeRepository.GetByIdAsync(employeeRowId);

            if (employee != null)
            {
                {
                    var withBlock = employee;
                    withBlock.LeaveBalance = 0;
                    withBlock.SickLeaveBalance = 0;
                }
            }
        }

        private LeaveTransaction CreateLeaveTransaction(PayrollContext context, LeaveTransactionType leaveTransactionType, LeaveLedger leaveLedger, PayPeriod payPeriod, decimal unusedLeaveHours)
        {
            var employeeRowId = leaveLedger.EmployeeID;

            var lt = context.LeaveTransactions.
                            Where(lTrans => lTrans.OrganizationID == _organizationId).
                            Where(lTrans => lTrans.EmployeeID == employeeRowId).
                            Where(lTrans => lTrans.LeaveLedgerID == leaveLedger.RowID).
                            Where(lTrans => lTrans.PayPeriodID == payPeriod.RowID).
                            Where(lTrans => lTrans.Type == leaveTransactionType.ToString()).
                            Where(lTrans => lTrans.Balance == 0).
                            Where(lTrans => lTrans.Amount == unusedLeaveHours).
                            Where(lTrans => lTrans.TransactionDate == payPeriod.PayToDate).
                            FirstOrDefault();

            if (lt == null)
            {
                lt = new LeaveTransaction()
                {
                    OrganizationID = _organizationId,
                    Created = DateTime.Now,
                    CreatedBy = _userId,
                    LastUpd = DateTime.Now,
                    LastUpdBy = _userId,
                    EmployeeID = employeeRowId,
                    LeaveLedgerID = leaveLedger.RowID,
                    PayPeriodID = payPeriod.RowID,
                    ReferenceID = null,
                    TransactionDate = payPeriod.PayToDate,
                    Type = leaveTransactionType.ToString(),
                    Balance = 0,
                    Amount = unusedLeaveHours
                };

                context.LeaveTransactions.Add(lt);

                ZeroOutEmployeeLeaveBalance(employeeRowId.Value);
            }

            return lt;
        }

        internal enum LeaveTransactionType
        {
            Credit,
            Debit
        }

        internal enum LeaveType
        {
            Vacation,
            Sick,
            All
        }

        internal enum AmountTreatment
        {
            Adjustment,
            Gross
        }
    }
}